### PR TITLE
Fix up ps quoting escape rules

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
When escaping quotes for a PowerShell argument we need to double up on more than just `'`. We also need to escape `'`, `’`, `‘`, `‚`, and `‛` by just repeating the char. Also fixes a bug when trying to escape `"` in a cmd argument where the escaped value need to be `\^"` and not just `^"`.

Fixes https://github.com/ansible-collections/ansible.windows/issues/45

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
quote

### CODE REVIEW

**Comment by jborean93** [comment]
Thanks @briantist for the pickup on this one. This should fix up the issue as well as fix a few encoding/cmd escaping edge cases I didn't test properly originally.

File: plugins/filter/quote.py [hunk 1]
```diff
@@ -8,15 +8,21 @@
...
 
-_UNSAFE_C = re.compile(r'[\s\t"]')
-_UNSAFE_CMD = re.compile(r'[\s\(\)\%\!^\"\<\>\&\|]')
+_UNSAFE_C = re.compile(u'[\\s\t"]')
+_UNSAFE_CMD = re.compile(u'[\\s\\(\\)\\^\\|%!"<>&]')
+
+# PowerShell has 5 characters it uses as a single quote, we need to double up on all of them.
+# https://github.com/PowerShell/PowerShell/blob/b7cb335f03fe2992d0cbd61699de9d9aafa1d7c1/src/System.Management.Automation/engine/parser/CharTraits.cs#L265-L272
+# https://github.com/PowerShell/PowerShell/blob/b7cb335f03fe2992d0cbd61699de9d9aafa1d7c1/src/System.Management.Automation/engine/parser/CharTraits.cs#L18-L21
+_UNSAFE_PWSH = re.compile(u"(['\u2018\u2019\u201a\u201b])")
```
Comment (briantist) [inline plugins/filter/quote.py:19]: I might have missed it but doesn't this list also need to include the "standard" single quote/apostrophe `'` along with the extended 4?

File: plugins/filter/quote.py [hunk 2]
```diff
@@ -46,21 +52,22 @@ def _quote_cmd(s):
...
     # https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_quoting_rules?view=powershell-5.1
     if not s:
-        return "''"
+        return u"''"
 
     # We should always quote values in PowerShell as it has conflicting rules where strings can and can't be quoted.
-    # This means we quote the entire arg with single quotes and escape the single quotes already in the arg.
-    return "'{0}'".format(s.replace("'", "''"))
+    # This means we quote the entire arg with single quotes and just double up on the single quote equivalent chars.
+    return u"'{0}'".format(_UNSAFE_PWSH.sub(u'\\1\\1', s))
```
Comment (briantist) [inline plugins/filter/quote.py:70]: Referring to my comment above, the character class would need single quote `'` so the sub catches it no?

File: plugins/filter/quote.py (hunk 1 above)
Comment (jborean93) [reply plugins/filter/quote.py:19]: Yep, that's the first character in the regex group, after the `[`.

Comment (briantist) [reply plugins/filter/quote.py:19]: ha I swear I looked it over 15 times before posting and yeah my eyes just went right over it. I figured it must have been something like that since all the tests pass.

**Comment by briantist** [comment]
#45 will be closed by this PR but lists at least 2 other instances like this in ansible/ansible. I can open sister issues there if you want.

**Comment by jborean93** [comment]
Up to you, once this was in I was going to just open a PR there with these fixes

**Comment by briantist** [comment]
> Up to you, once this was in I was going to just open a PR there with these fixes
works for me, no need to open a new one :)

**Comment by jborean93** [comment]
ansible/ansible PR opened at https://github.com/ansible/ansible/pull/68968.
